packages/cdktf/lib/terraform-stack.ts

```diff
@@ -28,6 +33,39 @@ export class TerraformStack extends Construct {
     super(scope, id);
 
     this.cdktfVersion = Node.of(this).tryGetContext("cdktfVersion");
+    Node.of(this).addValidation({
+      validate: () => {
+        const errors = [];
+        const tfConfig = this.toTerraform();
+
+        // validate provider has been initialized
+        const requiredProviders = [
+          ...new Set(
+            Object.keys({ ...tfConfig?.resource, ...tfConfig?.data }).map(
+              (providerName) => providerName.split("_")[0]
+            )
+          ),
+        ];
+
+        const providers = Object.keys(tfConfig?.provider || {});
+
+        const uninitializedProviders = requiredProviders.filter(
+          (reqProvider) =>
+            !providers.includes(reqProvider) && reqProvider !== "terraform" // e.g. terraform_remote_state
+        );
```

Comment:
> Unlike many other objects in the Terraform language, a provider block may be omitted if its contents would otherwise be empty.
[docs](https://www.terraform.io/docs/language/providers/configuration.html#provider-configuration-1)

Could we somehow check whether a provider that is not defined, has required attributes at all? That way we wouldn't force users to add empty provider declarations if they don't want to supply any config to them.

Reply:
i think you need them in the CDK though, #473 is with docker that does not need a provider config and still a problem as far as I understand 🤔

Reply:
Just checked, you're right. It seems to still be required.